### ISSUE TITLE
Go gin variable in handler is not passing to template

### ISSUE THREAD

**Issue by tomriddle1234** [gin-gonic/gin#1512]
- gin version (or commit ref): 1.3
- git version:
- operating system: ubuntu 18.04

## Description

https://stackoverflow.com/questions/52008320/go-gin-variable-in-handler-is-not-passing-to-template

**Comment by isgj** [comment]
From the code you have posted 
```go
 render(c, gin.H{"title": "Successful Login"}, "login-successful.html")
```
you are not passing the `is_logged_in`, so the second section `if not ...` will render.

**Comment by isgj** [comment]
No, the template can see only what you pass to the `c.HTML(...)` [(see method)](https://github.com/gin-gonic/gin/blob/6c6d97ba2e7eeaf78e20fc4344b21c4e057e2a9a/render/html.go#L69). What you set with `c.Set(...)` can be seen only by the middleware and the handler.

**Comment by GwynethLlewelyn** [comment]
Piping in late, but, besides thanking @isgj for explaining how Gin actually works and deals with these variables, I would suggest that this is turned into a feature request, for one simple reason: if you have just one or two templates it's easy to add all those extra variables; once you start developing a very complex site with lots of different templates, called from different places, then it's easy to miss one or two `c.HTML(...)` calls...
I _tried_ to add middleware to automatically add whatever variables I had to all router calls automatically, but, alas, this doesn't work, either (for precisely the same reason).

**Comment by barats** [comment]
Is there any resolution for this issue yet?

**Comment by GwynethLlewelyn** [comment]
@barats I'd say this is not really an issue any longer. You can see from the earlier examples (or from the enclosed StackOverflow link) that this error comes from either...

1. The template is not being loaded (on the quoted example on StackOverflow, nothing seems to call the template...).
2. `.Set()`, by design (and not bug!) just passes those values to the _middleware_, not to the template itself.

(and after my earlier post, I _did_ in fact figure out how to write a very small chunk of middleware and use it to set/reset common variables in my code, so that they get correctly passed to the template. It was easier than I thought!

**Comment by barats** [comment]
> @barats I'd say this is not really an issue any longer. You can see from the earlier examples (or from the enclosed StackOverflow link) that this error comes from either...
> 
> 1. The template is not being loaded (on the quoted example on StackOverflow, nothing seems to call the template...).
> 2. `.Set()`, by design (and not bug!) just passes those values to the _middleware_, not to the template itself.
> 
> (and after my earlier post, I _did_ in fact figure out how to write a very small chunk of middleware and use it to set/reset common variables in my code, so that they get correctly passed to the template. It was easier than I thought!

Sadly, you'll have to `.Get()` EVERY SINGLE variable in controller and pass it to EVERY SINGLE template, isn't it?
I don't think this is the correct way to do it.

**Comment by PatchRequest** [comment]
So if i have a middleware which sets a variable which is used in every template i have to extract it out of the context and pass it to the template in every function explicit? Adding a way to pass directoy from middleware to template would be a way to reduce redundand code
